### PR TITLE
6654_reactive_streams_concurrent_work

### PR DESCRIPTION
**Pull request by hutchig**
Signed-off-by: Gordon Hutchison <gordon.hutchison@gmail.com>
Concurrent tests for RSO

### CODE REVIEW

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
fixes #6654

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

File: dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/fat/src/com/ibm/ws/microprofile/reactive/streams/test/suite/ReactiveConcurrentWorkTest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                HttpUtils.findStringInReadyUrl(server, "/ReactiveConcurrentWorkTest/parallel?message=parallel" + count, FATServletClient.SUCCESS);
+                String msg = "Run on thread " + Thread.currentThread().getName();
+                System.out.println(msg);
+                return msg;
+            };
+            tasks.add(callableTask);
+        }
+
+        try {
+            List<Future<String>> futures = executor.invokeAll(tasks);
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/fat/src/com/ibm/ws/microprofile/reactive/streams/test/suite/ReactiveConcurrentWorkTest.java]: I think you need to wait for these futures to complete (or timeout) before shutting down the executor and ending the test.

File: dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/test-applications/ConcurrentWorkTest/src/com/ibm/ws/microprofile/reactive/streams/test/concurrent/ReactiveParallelClientTest.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * This method can be used to simulate a small stream run
+     *
+     * @param rsProcessorsInStream    the length of the pipeline is altered by adding this many Processors
+     * @param dataItemsSentDownStream once the stream is running this many onData's will be sent down it before it completes
+     */
+    public <T> void singleStreamWorkload(int rsProcessorsInStream, int dataItemsSentDownStream) {
+
+        // start of stream
+        PublisherBuilder<Integer> streamHead = ReactiveStreams.fromIterable(() -> {
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/test-applications/ConcurrentWorkTest/src/com/ibm/ws/microprofile/reactive/streams/test/concurrent/ReactiveParallelClientTest.java]: The operation of these streams is complicated so might benefit from reworking with a few more comments.

File: dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/test-applications/ConcurrentWorkTest/src/com/ibm/ws/microprofile/reactive/streams/test/concurrent/ReactiveParallelInServer.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    @Resource
+    private ManagedExecutorService mes; //Injected by the container
+
+    @GET
+    public String restEndpoint(@QueryParam("message") String message) {
+
+        System.out.println(message + " received on thread " + Thread.currentThread().getName());
+
+        // Set up the parallel tasks we create 3 * those that can be run on a physical core
+        int count = 3 * Runtime.getRuntime().availableProcessors();
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/test-applications/ConcurrentWorkTest/src/com/ibm/ws/microprofile/reactive/streams/test/concurrent/ReactiveParallelInServer.java]: is a machine dependent variable really needed for this test? ideally tests should run identically on any machine.

File: dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/fat/src/com/ibm/ws/microprofile/reactive/streams/test/suite/ReactiveConcurrentWorkTest.java (hunk 1 above)
Comment (hutchig) [reply dev/com.ibm.ws.microprofile.reactive.streams.operators_fat/fat/src/com/ibm/ws/microprofile/reactive/streams/test/suite/ReactiveConcurrentWorkTest.java]: OK I have inserted a wait for each (with a 5 min timeout)

**Comment by hutchig** [comment]
#build

**Pull request merged**